sklearn/cluster/_k_means_fast.pyx

```diff
@@ -7,155 +7,256 @@
 #         Lars Buitinck
 #
 # License: BSD 3 clause
-#
-# cython: boundscheck=False, wraparound=False, cdivision=True
 
-from libc.math cimport sqrt
 import numpy as np
-import scipy.sparse as sp
 cimport numpy as np
 cimport cython
 from cython cimport floating
+from libc.math cimport sqrt
+
+from ..utils.extmath import row_norms
+
+
+np.import_array()
 
-from ..utils.sparsefuncs_fast import assign_rows_csr
-from ..utils._cython_blas cimport _dot
 
 ctypedef np.float64_t DOUBLE
 ctypedef np.int32_t INT
 
 
-np.import_array()
+cdef floating _euclidean_dense_dense(floating* a,
+                                     floating* b,
+                                     int n_features,
+                                     bint squared) nogil:
+    """Euclidean distance between a dense and b dense"""
+    cdef:
+        int i
+        int n = n_features // 4
+        int rem = n_features % 4
+        floating result = 0
+
+    # We manually unroll the loop for better cache optimization.
+    for i in range(n):
+        result += ((a[0] - b[0]) * (a[0] - b[0])
+                  +(a[1] - b[1]) * (a[1] - b[1])
+                  +(a[2] - b[2]) * (a[2] - b[2])
+                  +(a[3] - b[3]) * (a[3] - b[3]))
+        a += 4; b += 4
+
+    for i in range(rem):
+        result += (a[i] - b[i]) * (a[i] - b[i])
+
+    return result if squared else sqrt(result)
+
+
+cdef floating _euclidean_sparse_dense(floating[::1] a_data,
```

Comment:
Maybe we could move _euclidean_dense_dense`, `_euclidean_sparse_dense`,  `_inertia_dense` and `_inertia_sparse` to `metrics/_pairwise_fast.pyx`? There is already `_sparse_manhattan` there that does similar things, as otherwise none will find them in this file when needed.

Reply:
Hum I tried to do that but I can't manage to compile then...
```
Exception: This may never happen, please report a bug
```
I'll dig into that later but it might involve fixing something in cython. I let it as is for now.